lib/Conversion/CalyxToFSM/MaterializeFSM.cpp

```diff
@@ -0,0 +1,199 @@
+//===- MaterializeCalyxToFSM.cpp - FSM Materialization Pass -----*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Contains the definitions of the FSM materialization pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "../PassDetail.h"
+#include "circt/Conversion/CalyxToFSM.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/FSM/FSMGraph.h"
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/OperationSupport.h"
+#include "llvm/ADT/STLExtras.h"
+
+using namespace circt;
+using namespace calyx;
+using namespace mlir;
+using namespace fsm;
+
+namespace {
+
+// A set of group names. Use SetVector to ensure deterministic ordering.
+using GroupSet = SetVector<StringAttr>;
+
+struct MaterializeCalyxToFSMPass
+    : public MaterializeCalyxToFSMBase<MaterializeCalyxToFSMPass> {
+  void runOnOperation() override;
+
+  struct StateEnableInfo {
+    llvm::SmallVector<Value> outputOperands;
+    llvm::SmallVector<Value> doneGuards;
+  };
+
+  /// Returns an operand vector for the states which should be enabled for a
+  /// given state, as well as the state done signals that should be considered
+  /// for the state transition guard. We do this here to avoid
+  /// assignStateOutputOperands/assignStateTransitionGuard having to each walk
+  /// the set of output operations.
+  StateEnableInfo gatherStateEnableInfo(StateOp stateOp) {
+    StateEnableInfo info;
+    size_t portIndex = 0;
+    auto &enabledGroups = stateEnables[stateOp];
+    for (auto group : referencedGroups) {
+      if (enabledGroups.contains(group)) {
+        info.outputOperands.push_back(c1);
+        info.doneGuards.push_back(machineOp.getArgument(portIndex));
+      } else
+        info.outputOperands.push_back(c0);
+
+      ++portIndex;
+    }
+    return info;
+  }
+
+  /// Assigns the 'fsm.output' operation of the provided 'state' to enabled the
+  /// set of provided groups. If 'topLevelDone' is set, also asserts the
+  /// top-level done signal.
+  void assignStateOutputOperands(StateOp stateOp,
+                                 llvm::SmallVector<Value> outputOperands,
```

Comment:
If I'm understanding correctly, the intention here is to make a copy and then append a value. This seems a bit tricky and I wonder if something more explicit would be better.

Reply:
> make a copy

What are you referring to here? (copy of what?)


The intention is to in-place modify the operands of the `fsm.output` operation within the `output` region of the state based on the set of operands provided.

I agree that the implementation looks a bit strange given that `gatherStateEnableInfo` is where the operands are defined. Opted to do it this way (as described by the `gatherStateEnableInfo` comment) to avoid having to walk over the set of referenced groups for a state.

For the sake of readability, this is probably premature optimization - i think I'll go ahead sink `gatherStateEnableInfo`  into `assignStateOutputOperands/assignStateTransitionGuard` to make them more readable.

Reply:
I mean outputOperands is passed by value and thus will make a copy

Reply:
Ahh, it makes sense now. Nevertheless, this has been refactored in the latest commit.